File: backend/src/models/Service.js
'use strict';

const mongoose = require('mongoose');

const serviceSchema = new mongoose.Schema(
  {
    centerId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'ServiceCenter',
      required: [true, 'Service center reference is required'],
    },
    name: {
      type: String,
      required: [true, 'Service name is required'],
      trim: true,
      maxlength: [100, 'Name must not exceed 100 characters'],
    },
    // Single uppercase letter used as token prefix (e.g. 'A' → A-001, A-002...)
    tokenPrefix: {
      type: String,
      required: [true, 'Token prefix is required'],
      uppercase: true,
      trim: true,
      maxlength: [3, 'Token prefix must not exceed 3 characters'],
    },
    description: {
      type: String,
      trim: true,
      maxlength: [300, 'Description must not exceed 300 characters'],
    },
    // Average time in minutes to serve one customer at this service
    avgServiceTimeMinutes: {
      type: Number,
      default: 8,
      min: [1, 'Average service time must be at least 1 minute'],
    },
    isActive: {
      type: Boolean,
      default: true,
    },
    order: {
      type: Number,
      default: 0, // Display order
    },
  },
  {
    timestamps: true,
  }
);

// ─── Indexes ──────────────────────────────────────
serviceSchema.index({ centerId: 1 });
serviceSchema.index({ centerId: 1, isActive: 1 });
// Ensure token prefix is unique within a center
serviceSchema.index({ centerId: 1, tokenPrefix: 1 }, { unique: true });

const Service = mongoose.model('Service', serviceSchema);

module.exports = Service;

File: user_app/lib/models/service.dart
class Service {
  const Service({
    required this.id,
    required this.name,
    required this.code,
    required this.centerId,
    required this.isActive,
    this.prefix,
    this.estimatedDuration,
    this.description,
    this.priorityAllowed = false,
  });

  final String id;
  final String name;
  final String code;
  final String centerId;
  final bool isActive;
  final String? prefix;
  final int? estimatedDuration;
  final String? description;
  final bool priorityAllowed;

  factory Service.fromJson(Map<String, dynamic> json) {
    return Service(
      id: (json['_id'] ?? json['id'] ?? '').toString(),
      name: (json['name'] ?? '').toString(),
      code: (json['code'] ?? '').toString(),
      centerId: (json['centerId'] is Map ? json['centerId']['_id'] : json['centerId'] ?? '').toString(),
      isActive: json['isActive'] == true,
      prefix: json['prefix']?.toString(),
      estimatedDuration: (json['estimatedDuration'] as num?)?.toInt(),
      description: json['description']?.toString(),
      priorityAllowed: json['priorityAllowed'] == true,
    );
  }
}
